File: worker/src/guardrails.ts
/**
 * Guardrail Layer B — stops scripted abuse from exhausting the free model
 * quota or running up a bill. Two independent checks; both must pass:
 *
 *   1. Turnstile: proves the caller executed real browser JS and solved
 *      Cloudflare's challenge. Blocks curl/script traffic outright.
 *   2. Rate limit (KV): caps requests per IP even from a real browser, in
 *      case a token is reused or a human hammers the endpoint.
 */

export interface GuardrailEnv {
  RATE_LIMIT: KVNamespace
  TURNSTILE_SECRET_KEY?: string
}

const RATE_LIMIT_PER_MINUTE = 10
const RATE_LIMIT_PER_DAY = 50

export async function verifyTurnstile(secretKey: string, token: string, ip: string): Promise<boolean> {
  try {
    const res = await fetch('https://challenges.cloudflare.com/turnstile/v0/siteverify', {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: new URLSearchParams({ secret: secretKey, response: token, remoteip: ip }),
    })
    const data = (await res.json()) as { success?: boolean }
    return data.success === true
  } catch {
    return false
  }
}

/**
 * Fixed-window counters in KV. Not perfectly atomic (a concurrent get-then-put
 * race can under-count by a handful of requests), which is fine here — this
 * is an abuse backstop, not a billing meter.
 */
export async function checkRateLimit(env: GuardrailEnv, ip: string): Promise<boolean> {
  const now = new Date()
  const minuteBucket = Math.floor(now.getTime() / 60_000)
  const dayBucket = now.toISOString().slice(0, 10)
  const minuteKey = `rl:min:${ip}:${minuteBucket}`
  const dayKey = `rl:day:${ip}:${dayBucket}`

  const [minuteRaw, dayRaw] = await Promise.all([
    env.RATE_LIMIT.get(minuteKey),
    env.RATE_LIMIT.get(dayKey),
  ])
  const minuteCount = minuteRaw ? parseInt(minuteRaw, 10) : 0
  const dayCount = dayRaw ? parseInt(dayRaw, 10) : 0

  if (minuteCount >= RATE_LIMIT_PER_MINUTE || dayCount >= RATE_LIMIT_PER_DAY) {
    return false
  }

  await Promise.all([
    env.RATE_LIMIT.put(minuteKey, String(minuteCount + 1), { expirationTtl: 60 }),
    env.RATE_LIMIT.put(dayKey, String(dayCount + 1), { expirationTtl: 60 * 60 * 24 }),
  ])
  return true
}

File: aws-bot/server/app/guardrails.py
"""Guardrail Layer B — same two checks as worker/src/guardrails.ts, ported.

1. Turnstile: proves the caller ran real browser JS, blocks scripted traffic.
2. Rate limit (DynamoDB, atomic per-counter via conditional UpdateItem —
   stricter than the Worker's KV version, which admits a get-then-put race).
"""

import os
import time
import urllib.parse
import urllib.request

import boto3
from botocore.exceptions import ClientError

RATE_LIMIT_PER_MINUTE = 4
RATE_LIMIT_PER_DAY = 15

TABLE_NAME = os.environ.get("RATE_LIMIT_TABLE")
TURNSTILE_SECRET_KEY = os.environ.get("TURNSTILE_SECRET_KEY")

_table = None


def _get_table():
    global _table
    if _table is None:
        region = os.environ.get("AWS_REGION_NAME", "eu-west-1")
        _table = boto3.resource("dynamodb", region_name=region).Table(TABLE_NAME)
    return _table


def verify_turnstile(token: str, ip: str) -> bool:
    if not TURNSTILE_SECRET_KEY:
        return False
    data = urllib.parse.urlencode(
        {"secret": TURNSTILE_SECRET_KEY, "response": token, "remoteip": ip}
    ).encode()
    req = urllib.request.Request(
        "https://challenges.cloudflare.com/turnstile/v0/siteverify",
        data=data,
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=5) as res:
            import json

            return json.loads(res.read()).get("success") is True
    except Exception:
        return False


def _increment_counter(pk: str, limit: int, ttl_seconds: int) -> bool:
    """Atomically increments a fixed-window counter. Returns False if it's at the limit."""
    try:
        _get_table().update_item(
            Key={"pk": pk},
            UpdateExpression="ADD hits :one SET #ttl = if_not_exists(#ttl, :ttl)",
            ConditionExpression="attribute_not_exists(hits) OR hits < :limit",
            ExpressionAttributeNames={"#ttl": "ttl"},
            ExpressionAttributeValues={
                ":one": 1,
                ":limit": limit,
                ":ttl": int(time.time()) + ttl_seconds,
            },
        )
        return True
    except ClientError as e:
        if e.response["Error"]["Code"] == "ConditionalCheckFailedException":
            return False
        raise


def check_rate_limit(ip: str) -> bool:
    if not TABLE_NAME:
        return False
    now = int(time.time())
    minute_bucket = now // 60
    day_bucket = now // 86400

    if not _increment_counter(f"min#{ip}#{minute_bucket}", RATE_LIMIT_PER_MINUTE, 60):
        return False
    if not _increment_counter(f"day#{ip}#{day_bucket}", RATE_LIMIT_PER_DAY, 86400):
        return False
    return True
